src/net.cpp

```diff
@@ -718,6 +720,127 @@ CNetMessage V1TransportDeserializer::GetMessage(const CMessageHeader::MessageSta
     return msg;
 }
 
+int V2TransportDeserializer::Read(const char* pch, unsigned int bytes)
+{
+    if (!m_in_data) {
+        // copy data to temporary parsing buffer
+        const unsigned int remaining = CHACHA20_POLY1305_AEAD_AAD_LEN - m_hdr_pos;
+        const unsigned int copy_bytes = std::min(remaining, bytes);
+
+        memcpy(&vRecv[m_hdr_pos], pch, copy_bytes);
+        m_hdr_pos += copy_bytes;
+
+        // if AAD incomplete, exit
+        if (m_hdr_pos < CHACHA20_POLY1305_AEAD_AAD_LEN) {
+            return copy_bytes;
+        }
+
+        // we got the AAD bytes at this point (3 bytes encrypted packet length)
+        // we keep the sequence numbers unchanged at this point. Once the message is authenticated and decrypted, we increase the sequence numbers (or the aad_pos)
+        if (!m_aead->GetLength(&m_message_size, m_aad_seqnr, m_aad_pos, (const uint8_t*)vRecv.data())) {
+            return -1;
+        }
+
+        // reject messages larger than MAX_SIZE
+        if (m_message_size > MAX_SIZE) {
+            return -1;
+        }
+
+        // switch state to reading message data
+        m_in_data = true;
+
+        return copy_bytes;
+    } else {
+        const unsigned int remaining = m_message_size + CHACHA20_POLY1305_AEAD_TAG_LEN - m_data_pos;
+        const unsigned int copy_bytes = std::min(remaining, bytes);
+
+        // extend buffer, respect previous copied AAD part
+        if (vRecv.size() < CHACHA20_POLY1305_AEAD_AAD_LEN + m_data_pos + copy_bytes) {
+            // Allocate up to 256 KiB ahead, but never more than the total message size (incl. AAD & TAG).
+            vRecv.resize(std::min(CHACHA20_POLY1305_AEAD_AAD_LEN + m_message_size + CHACHA20_POLY1305_AEAD_TAG_LEN, CHACHA20_POLY1305_AEAD_AAD_LEN + m_data_pos + copy_bytes + 256 * 1024 + CHACHA20_POLY1305_AEAD_TAG_LEN));
+        }
+
+        memcpy(&vRecv[CHACHA20_POLY1305_AEAD_AAD_LEN + m_data_pos], pch, copy_bytes);
+        m_data_pos += copy_bytes;
+
+        return copy_bytes;
+    }
+}
+
+CNetMessage V2TransportDeserializer::GetMessage(const CMessageHeader::MessageStartChars& message_start, int64_t time)
+{
+    // In v2, vRecv contains the encrypted payload plus the MAC tag (1+bytes serialized message command + ? bytes message payload + 16 byte mac tag)
+    assert(Complete());
+
+    // defensive decoding (MAC check, decryption, command deserialization)
+    // we'll always return a CNetMessage (even if encryption fails), we always increase the AEAD sequence numbers
```

Comment:
Even if decryption fails ? Also comment about increasing AAD sequence number is allusive can you precise a unsuccessful decryption still increase sequence number.

Reply:
What is the problem of increasing the sequence number if decryption fails? We detect it and can respond (disconnecting) on the layer above the transport.

Reply:
Right, decryption failure is fatal, there is no such ban-per-point in this case.